pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +380,673 @@ pub mod pallet {
+						let remaining_foreign_amount = T::CurrencyConverter::stable_to_stable(
+							info.base.foreign_currency,
+							Pallet::<T>::pool_currency_of(investment_id)?,
+							T::Investment::investment(&who, investment_id)?,
+						)?;
+
+						T::DecreasedForeignInvestOrderHook::notify_status_change(
+							(who.clone(), investment_id),
+							ExecutedForeignDecreaseInvest {
+								amount_decreased: info.decrease_swapped_amount,
+								foreign_currency: info.base.foreign_currency,
+								amount_remaining: remaining_foreign_amount,
+							},
+						)?;
+
+						info.decrease_swapped_amount = T::Balance::default();
+
+						if info.remaining_pool_amount()?.is_zero() {
+							*maybe_info = None;
+						}
+					}
+				}
+
+				Ok(())
+			})
+		}
+
+		/// Notifies that a partial swap has been done and applies the result to
+		/// an `RedemptionInfo`
+		fn notify_redemption_swap_done(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			swapped_amount: T::Balance,
+			pending_amount: T::Balance,
+		) -> DispatchResult {
+			ForeignRedemptionInfo::<T>::mutate_exists(&who, investment_id, |maybe_info| {
+				let info = maybe_info.as_mut().ok_or(Error::<T>::InfoNotFound)?;
+
+				info.swapped_amount.ensure_add_assign(swapped_amount)?;
+				if pending_amount.is_zero() {
+					let redemption = T::Investment::redemption(&who, investment_id)?;
+
+					T::CollectedForeignRedemptionHook::notify_status_change(
+						(who.clone(), investment_id),
+						ExecutedForeignCollect {
+							currency: info.base.foreign_currency,
+							amount_currency_payout: info.swapped_amount,
+							amount_tranche_tokens_payout: info.collected_tranche_tokens(),
+							amount_remaining: redemption,
+						},
+					)?;
+
+					info.pending_tranche_tokens
+						.ensure_sub_assign(info.collected_tranche_tokens())?;
+
+					info.base.collected = CollectedAmount::default();
+					info.swapped_amount = T::Balance::default();
+
+					if info.pending_tranche_tokens.is_zero() {
+						*maybe_info = None;
+					}
+				}
+
+				Ok(())
+			})
+		}
+	}
+
+	impl<T: Config> ForeignInvestment<T::AccountId> for Pallet<T> {
+		type Amount = T::Balance;
+		type CurrencyId = T::CurrencyId;
+		type Error = DispatchError;
+		type InvestmentId = T::InvestmentId;
+
+		fn increase_foreign_investment(
+			who: &T::AccountId,
+			investment_id: T::InvestmentId,
+			foreign_amount: T::Balance,
+			foreign_currency: T::CurrencyId,
+		) -> DispatchResult {
+			// NOTE: This line will be removed with market ratios
+			let pool_amount = T::CurrencyConverter::stable_to_stable(
+				Self::pool_currency_of(investment_id)?,
+				foreign_currency,
+				foreign_amount,
+			)?;
+
+			ForeignInvestmentInfo::<T>::mutate(&who, investment_id, |info| {
+				let info = info.get_or_insert(InvestmentInfo::new(foreign_currency)?);
+
+				info.base.ensure_same_foreign(foreign_currency)?;
+				info.total_pool_amount.ensure_add_assign(pool_amount)?;
+
+				Ok::<_, DispatchError>(())
+			})?;
```

Comment:
couldn't we make the `foreign_currency` part of the key? Easily allow to invest with multiple currencies in paralel while ommitting the check and storage of the `foreign_currency` alltogether.

Reply:
Not trivial, the `StatusNotificationHook` from collected investment and redemption does not identify with a foreign currency to know how to proceed.

Reply:
I would start this as a new feature for later